modules/_labs/select/react/lib/Select.tsx

```diff
@@ -309,39 +358,57 @@ export default class Select extends React.Component<SelectProps, SelectState> {
 
   componentWillUnmount() {
     // Clear timers
-    if (this.removeMenuTimer) {
-      clearTimeout(this.removeMenuTimer);
+    if (this.menuAnimationTimer) {
+      clearTimeout(this.menuAnimationTimer);
     }
     if (this.clearKeysSoFarTimer) {
       clearTimeout(this.clearKeysSoFarTimer);
     }
   }
 
-  handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
-    const {isMenuHidden} = this.state;
-
-    // Return focus to the button if the menu is visible (i.e., we're
-    // clicking the button again to dismiss the menu)
-    if (!isMenuHidden) {
-      this.focusButton();
-    }
-
-    this.toggleMenu(isMenuHidden);
-  };
-
   handleMouseDown = (event: React.MouseEvent<HTMLButtonElement>): void => {
     // Cancel default handling of the event to ensure we maintain control
     // of focus (i.e., so focus is immediately transferred to the menu when
     // opening the menu, rather than briefly being applied to the button
-    // and creating a momentary flash of a focus ring around the button)
+    // and creating a momentary flash of a focus ring around the button).
+    // This has implications for our click handler (see handleClick below).
     event.preventDefault();
   };
 
+  handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
+    const {menuVisibility} = this.state;
+
+    switch (menuVisibility) {
+      // If we click the button while the menu is in the process of closing,
+      // we want to toggle the menu back on. However, we also need to focus
+      // the menu since it won't be focused using Popper's onFirstUpdate
+      // callback (because the menu already exists). If we don't focus the
+      // menu, clicking outside the menu to dismiss it on blur won't work.
+      case 'close':
+      case 'closing':
+        if (this.menuRef.current) {
+          this.menuRef.current.focus();
+        }
+        this.toggleMenu(true);
+        break;
+      case 'closed':
+        this.toggleMenu(true);
+        break;
+      // Otherwise, the menu is opened or in the process of opening. Focus
+      // the button and toggle the menu off. Note that since we're calling
+      // event.preventDefault in our mouseDown handler for the button, we
+      // must manage focus on the button ourselves (the browser will no
+      // longer automatically apply focus to the button when clicking on it).
+      default:
+        this.focusButton();
+        this.toggleMenu(false);
+        break;
+    }
+  };
+
   handleOptionSelection = (index: number): void => {
-    // Abort immediately if:
-    // * The menu is in the process of hiding
-    // * A disabled option was clicked (we ignore these clicks)
-    if (this.state.isMenuHiding || this.normalizedOptions[index].disabled) {
+    // Abort if a disabled option was clicked (we ignore these clicks)
+    if (this.normalizedOptions[index].disabled) {
       return;
     }
```

Comment:
Snuck in a quick fix to #872 (no longer aborts if the menu is in the process of closing). It's debatable if this check was really needed in the first place -- is it really a problem if an option is selected while the menu is in the middle of fading out?